XIPs/xip-46-multi-wallet-identity.md

```diff
@@ -0,0 +1,520 @@
+
+##### CreateInbox
+
+```text
+- Create inbox
+  (Owner: ${INITIAL_ADDRESS})
+```
+
+##### AddAssociation (Installation Key)
+
+```text
+- Grant messaging access to app
+  (ID: ${hex(INSTALLATION_PUBLIC_KEY)})
+```
+
+##### AddAssociation (Associated Address)
+
+```text
+- Link address to inbox
+  (Address: ${ASSOCIATED_ADDRESS})
+```
+
+##### RevokeAssociation (Installation Key)
+
+```text
+- Revoke messaging access from app
+  (ID: ${hex(INSTALLATION_PUBLIC_KEY)})
+```
+
+##### Revoke Association (Associated Address)
+
+```text
+- Unlink address from inbox
+  (Address: ${ASSOCIATED_ADDRESS})
+```
+
+##### ChangeRecoveryAddress
+
+```text
+- Change inbox recovery address
+  (Address: ${NEW_RECOVERY_ADDRESS})
+```
+
+Finally, a footer is appended.
+
+##### Footer
+
+```text
+
+For more info: https://xmtp.org/signatures/
+```
+
+The identity update will have one or more signers. If a signer is an installation key, this signing will be performed directly by the application within the libxmtp SDK, and the user will not see the signing text.
+
+If a signer is a wallet that the user controls, the signing text will be visible to the user and require their explicit acceptance.
+
+For example, an app may expose a UI to associate another wallet to the existing Inbox ID (in other words, an `AddAssociation`). Once the user initiates this action, the `existing_member_signature` will typically be signed by an installation key from the app, while the `new_member_signature` will be signed by the wallet. The latter signature will require the user to explicitly confirm the operation from the new wallet.
+
+#### Signature formats
+
+These identity updates accept the following signature formats to facilitate signatures from EOA wallets, smart contract wallets, and installation keys using the ECDSA signature format, as well as legacy XMTP v2 keys used for the migration of existing accounts.
+
+```protobuf
+// RecoverableEcdsaSignature for EIP-191 and v2 signatures
+message RecoverableEcdsaSignature {
+  // 65-bytes [ R || S || V ], with recovery id as the last byte
+  bytes bytes = 1;
+}
+
+// EdDSA signature for 25519
+message RecoverableEd25519Signature {
+  bytes bytes = 1;
+}
+
+// Smart wallet signature
+message Erc1271Signature {
+  // CAIP-10 contract address
+  // https://github.com/ChainAgnostic/CAIPs/blob/main/CAIPs/caip-10.md
+  string contract_address = 1;
+  // Specify the block height to verify the signature against
+  int64 block_height = 2;
+  // The actual signature bytes
+  bytes signature = 3;
+}
+
+// An existing address on xmtpv2 may have already signed a legacy identity key
+// of type SignedPublicKey via the 'Create Identity' signature.
+// For migration to xmtpv3, the legacy key is permitted to sign on behalf of the
+// address to create a matching xmtpv3 installation key.
+// This signature type can ONLY be used for CreateInbox and AddAssociation
+// payloads, and can only be used once in xmtpv3.
```

Comment:
Does "only once" mean you have to concatenate the CreateInbox and AddAssociation into a single IdentityUpdate?